src/gwt/src/org/rstudio/studio/client/workbench/ui/PaneManager.java

```diff
@@ -674,18 +675,51 @@ public void onFocusPreviousPane()
    }
 
    @Handler
-    public void onNewSourceColumn()
-    {
-       if (!userPrefs_.allowSourceColumns().getValue())
-          pGlobalDisplay_.get().showMessage(GlobalDisplay.MSG_INFO, "Cannot Add Column",
-             "Allow Source Columns preference is disabled.");
-       else if (additionalSourceCount_ >= MAX_COLUMN_COUNT)
-          pGlobalDisplay_.get().showMessage(GlobalDisplay.MSG_INFO, "Cannot Add Column",
-             "You can't add more than " + MAX_COLUMN_COUNT + " columns.");
-       else
-          createSourceColumn();
-    }
+   public void onNewSourceColumn()
+   {
+      if (validateNewColumnRequest())
+         createAndDisplaySourceColumn();
+   }
+   
+   @Handler
+   public void onOpenSourceDocNewColumn()
+   {
+      if (validateNewColumnRequest())
+      {
+         ColumnName name = createSourceColumn();
+         SourceColumn column = sourceColumnManager_.getByName(name.getName());
+         column.incrementNewTabPending();
+         panel_.addLeftWidget(
+            createSourceColumnWindow(name.getName(),
+               name.getAccessibleName()));
+         sourceColumnManager_.setActive(column);
+         Command onCancelled = () ->
+         {
+            column.decrementNewTabPending();
+            closeSourceWindow(name.getName());
+         };
+         Command onCompleted = () -> column.decrementNewTabPending();
+         source_.openSourceDoc(onCancelled, onCompleted);
+      }
+   }
 
+   private boolean validateNewColumnRequest()
+   {
+      if (!userPrefs_.allowSourceColumns().getValue())
+      {
+         pGlobalDisplay_.get().showMessage(GlobalDisplay.MSG_INFO, "Cannot Add Column",
+            "Allow Source Columns preference is disabled.");
```

Comment:
This command needs to be disabled, and probably hidden, when the `allowSourceColumns` preference prevents it from working. I'd suggest handling that in the source column manager (i.e. bind a preference listener that sets the command to be enabled/visible based on the value of the pref).

Reply:
Good point, I've added a listener to `SourceColumnManager` to handle this as well as updated PaneManager to hide the commands when the max column count is reached.